neo/Consensus/ConsensusService.cs

```diff
@@ -225,9 +286,281 @@ private void OnPersistCompleted(Block block)
+                        // Can't use recover if the we are the primary and we haven't sent a prepare request and no
+                        // prepare request was present in the recovery message.
+                        if (myIndexIsPrimary) return;
+
+                        // If we have no `Preparation` hash we can't regenerate any of the `PrepareRequest` messages.
+                        if (message.PreparationHash == null) return;
+                        preparationHash = message.PreparationHash;
+                    }
+                }
+
+                for (int i = 0; i < context.Validators.Length; i++)
+                {
+                    // If we are missing this preparation.
+                    if (context.Preparations[i] != null) continue;
+                    if (i == context.PrimaryIndex) continue;
+                    // If the recovery message has this preparations
+                    if (message.PrepareWitnessInvocationScripts[i] == null) continue;
+                    var prepareResponseMsg = new PrepareResponse {PreparationHash = preparationHash};
+                    prepareResponseMsg.ViewNumber = context.ViewNumber;
+                    var regeneratedPrepareResponse = ((ConsensusContext) context).RegenerateSignedPayload(
+                        prepareResponseMsg, (ushort) i, message.PrepareWitnessInvocationScripts[i],
+                        message.PrepareTimestamps[i]);
+                    if (regeneratedPrepareResponse.Verify(snap) &&
+                        PerformBasicConsensusPayloadPreChecks(regeneratedPrepareResponse))
+                        OnPrepareResponseReceived(regeneratedPrepareResponse, prepareResponseMsg);
+                }
+            }
+
+            if (!context.State.HasFlag(ConsensusState.CommitSent)) return;
+
+            RestoreCommits(message);
+        }
+
+        private void OnRecoveryMessageReceived(ConsensusPayload payload, RecoveryMessage message)
+        {
+            if (context.BlockIndex > payload.BlockIndex) return;
+            if (context.State.HasFlag(ConsensusState.BlockSent)) return;
+            Snapshot snap =  Blockchain.Singleton.GetSnapshot();
+            if (payload.BlockIndex > snap.Height + 1) return;
+            if (message.PrepareWitnessInvocationScripts.Length < context.Validators.Length) return;
+            if (message.PrepareTimestamps.Length < context.Validators.Length) return;
+
+            Log($"{nameof(OnRecoveryMessageReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex}");
+
+            if (context.ViewNumber == message.ViewNumber)
+            {
+                HandleRecoveryInCurrentView(message, snap);
+                return;
+            }
+
+            // Commited nodes cannot change view or it can lead to a potential block spork, since their block signature
+            // could potentially be used twice.
+            if (context.State.HasFlag(ConsensusState.CommitSent)) return;
```

Comment:
If the commit is sent it can still accept more commits in the recovery message if it sees a recovery message that has more commits. This can help it generate a block faster.

Reply:
It is actually required now to have it this way, because I now changed it so that when a node dies and restarts that has`CommitSent`, that node will periodically sent the `RecoveryMessage` instead of the `Commit` message. So it has to be this way so that they will be able to accept each others commits if they missed the messages.

Reply:
The idea looks good to me.